src/brevitas_examples/common/learned_round/learned_round_optimizer.py

```diff
@@ -508,25 +510,32 @@ def _optimize_learned_round_block(
         init_loss = -1.0
         last_best_iter = self.iters
 
+        scaler = None
+        use_amp = next(block.parameters()).dtype == torch.float32
+        if use_amp:
+            scaler = GradScaler()
+
         # Dictionary to store the rounding parameters yielding the lowest
         # training loss
         optimal_rounding_params = {}
-        torch.autograd.set_detect_anomaly(True)
         n_samples = len(cache)
         pbar = tqdm(range(self.iters), desc='')
         for i in pbar:
             # Sample mini-batch from cache
             idxs = torch.randperm(n_samples)[:self.batch_size]
             inputs, fp_outs = cache.sample_batch(idxs)
 
-            # Run block forward to obtain quant outputs
-            quant_outs = block_forward(block, inputs)
-            fp_outs = send_to_device(fp_outs, quant_outs.device)
-            if self.use_amp:
+            if use_amp:
                 with autocast(device_type="cuda" if torch.cuda.is_available() else "cpu",
                               dtype=self.amp_dtype):
+                    # Run block forward to obtain quant outputs
+                    quant_outs = block_forward(block, inputs)
+                    fp_outs = send_to_device(fp_outs, quant_outs.device)
                     loss, loss_components = block_loss(quant_outs, fp_outs)
             else:
+                # Run block forward to obtain quant outputs
+                quant_outs = block_forward(block, inputs)
+                fp_outs = send_to_device(fp_outs, quant_outs.device)
                 loss, loss_components = block_loss(quant_outs.to(torch.float32), fp_outs.to(torch.float32))
```

Comment:
The code for each condition is almost exactly the same. Maybe we could have `autocast(enabled=use_amp, ...)` and just have a conditional for upcasting the outputs to float32 before computing the loss to avoid repeting block_forward/send_to_device/block_loss.

Reply:
I understand it's less code, but I think it will be more confusing. I am leaving as it is. Extra verbosity for clarity, I am happy with that.